### PR TITLE
SHOR-44: Backstop Js test suite Integration

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview
PR contains Initial bootstrap code for backstopJS test suite and screens added via.
PRs 
* https://github.com/swastikpareek/org.civicrm.civicase/pull/1
* https://github.com/swastikpareek/org.civicrm.civicase/pull/2

### List of final screens covered.
* Civicase - Dashboard
* Civicase - Find Cases - Search
* Civicase - Find My Cases
* Civicase - Add Case
* Civicase - Manage Case
* Civicase - Delete Case
* Civicase - Assign to others case
* Civicase - View case activity
* Civicase - Case Report
* Civicase - New Case Report
* Civicase - Case Summary Report - Columns
* Civicase - Case Summary Report - Sorting
* Civicase - Case Summary Report - Filters
* Civicase - Case Time Spent Report - Grouping
* Civicase - Case Detail Report - Display Options
* Civicase - Constituent Summary Report - Title and Format
* Civicase - Constituent Summary Report - Email Delivery
* Civicase - Constituent Summary Report - Access
* Civicase - Settings
* Civicase - Settings - Case Type
* Civicase - Settings - New Case Type - Case Roles
* Civicase - Settings - New Case Type - Case Statusses
* Civicase - Settings - New Case Type - Activity Types
* Civicase - Settings - New Case Type - Standard Timelines
* Civicase - Settings - Redaction Rule Options
* Civicase - Settings - Redaction Rule Options - Add New Redaction Rule
* Civicase - Settings - Redaction Rule Options - Delete Redaction Rule
* Civicase - Settings - Case Status Options
* Civicase - Settings - Case Status Options - Edit
* Civicase - Settings - Case Status Options - Disable
* Civicase - Settings - Case Status Options - Delete
* Civicase - Settings - Encounter Medium Options
* Civicase - Settings - Encounter Medium Options - Edit

### Technical Details
* Gulp based test suite
* All scenarios should be created in `tests/backstop_data/backstop.tpl.jsom` folder.
* Screenshots captured are saved inside `backstop_data/screenshots` folder.
#### How to use it
1. Create a `site-config.json` file with the following content.
    ```json
    {
      "url": "your_local_url",
      "root": "absolute_path_to_site"
    }
    ```
2. Create the reference screenshots
    ```shell
    gulp backstopjs:reference
    ```
3. Create the test screenshots and compare
    ```shell
    gulp backstopjs:test
    ```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,3 +45,7 @@ civibuild create dcase --url http://dcase.localhost --admin-pass s3cr3t
 ## Create a build with WordPress, CiviCRM 4.7, and CiviCase 5
 civibuild create wpcase --url http://wpcase.localhost --admin-pass s3cr3t
 ```
+
+## Post Development
```
Comment (AkA84) [inline README.md]: Let's not change the README if we are not told so, we don't have a "Development" section so let's leave this file as-is

File: package.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "backstopjs-test",
+  "version": "1.0.0",
+  "description": "BackstopJS configuration for CiviCRM - Civicase",
+  "main": "index.js",
+  "scripts": {
+    "precommit": "lint-staged",
+    "test": "node test-runner.js"
+  },
+  "lint-staged": {
```
Comment (AkA84) [inline package.json]: this is a _huge_ change, you are introducing a linter that will block any commit that won't pass its checks.

This goes way beyond what the scope of this PR is, so either you make it very clear in the PR that you are introducing this and ask the owners of the extension if they're fine with it, or you don't do it at all

File: tests/backstop_data/backstop.tpl.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  "onBeforeScript": "onBefore.js",
+  "onReadyScript": "onReady.js",
+  "paths": {
+    "bitmaps_reference": "screenshots/reference",
+    "bitmaps_test": "screenshots/test",
+    "engine_scripts": "engine_scripts/puppet",
+    "html_report": "html_report",
+    "ci_report": "ci_report"
+  },
+  "scenarios": [{
```
Comment (AkA84) [inline tests/backstop_data/backstop.tpl.json]: leave the scenarios list empty

File: README.md (hunk 1 above)
Comment (swastikpareek) [reply README.md]: Done!

File: package.json (hunk 2 above)
Comment (swastikpareek) [reply package.json]: Done!

File: tests/backstop_data/backstop.tpl.json (hunk 3 above)
Comment (swastikpareek) [reply tests/backstop_data/backstop.tpl.json]: Done!

File: package.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "backstopjs-test",
```
Comment (AkA84) [inline package.json]: this is the package.json for the whole extension, not just for its backstop suite. Check how the file is in mosaico https://github.com/veda-consulting/uk.co.vedaconsulting.mosaico/blob/2.x/package.json

Comment (swastikpareek) [reply package.json]: Done!

**Comment by colemanw** [comment]
Do you have a sense of what it will take to add Github CI to this test suite so that tests are run on incoming PRs?

**Comment by AkA84** [comment]
@colemanw could you merge this please?

File: TESTING.md [hunk 5]
```diff
@@ -57,3 +57,37 @@ Time: 16.86 seconds, Memory: 52.25Mb
...
+    ```shell
+    gulp backstopjs:test
+    ```
+
+#### Parallel capturing
+BackstopJS supports taking multiple screenshot at once. Change the value of `asyncCaptureLimit` in _backstop.tpl.json_ to decide how many screenshots you want to take in parallel
+
+***Note**: Please be aware that BackstopJS performance is heavily dependent on the specs of the machine it runs on, so make sure to choose a value that the tool can handle on your machine (otherwise you will encounter random timeout errors)*
+
+***Note**: The backstop test suite for now, only covers screens that are exposed by civicase *component* (and not the extension) that comes by default with core. The backstop JS test suite can be extended to cover the screens that are exposed by the extension.
```
Comment (AkA84) [inline TESTING.md]: Let's simplify this:
```md
## BackstopJS Visual Regression testing
This test suite is based on [BackstopJS](https://garris.github.io/BackstopJS) plugin. Backstop JS uses pupetter and headless chrome to create reference screenshots and use them to compare new screenshots and raise any incosistency in the visuals of the page (if introduced while developing something)

***Note**: The backstop test suite for now only covers screens that are exposed by civicase *component* (and not the extension) that comes by default with core.

Documentation available [here](https://github.com/garris/BackstopJS#backstopjs)
```

File: tests/backstop_data/backstop.tpl.json [hunk 6]
```diff
@@ -0,0 +1,193 @@
+{
+  "id": "backstop_civicase",
```
Comment (AkA84) [inline tests/backstop_data/backstop.tpl.json]:
```json
"id": "civicase_legacy"
```

**Comment by swastikpareek** [comment]
Closing this PR as the screens covered under this are not part of the extension but the part of the civicrm core. The new PR will have these changes and will be against the core backstop JS test suite - https://github.com/compucorp/backstopjs-config.

**Pull request closed without merging**